### PR TITLE
Rewrite assembly identity must match tests to use in memory assemblies

### PR DESCRIPTION
**Pull request by safern**
While adding some rules and running tests I noticed that the new tests for assembly identity rule bumped the execution time from ~10s to ~30-40s in the console and to 1+min in VS. Also depending on machine state sometimes they failed when switching branches. 

This was causing tests to generate a project and build it to get some dlls out and then load them, which for testing purposes is not worth it as it brings some flakiness, we should test rules with in memory assemblies as we just want to test that the rules are interpreting the metadata correctly.

Also I did some cleanup on the rule and improved a diagnostic message.

### CODE REVIEW

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/AssemblyIdentityMustMatchTests.cs [hunk 1]
```diff
@@ -30,170 +36,106 @@ public static void AssemblyNamesDoNotMatch()
...
-            IAssemblySymbol leftSymbols = new AssemblySymbolLoader().LoadAssembly(leftDllPath);
-            IAssemblySymbol rightSymbols = new AssemblySymbolLoader().LoadAssembly(rightDllPath);
+            IAssemblySymbol leftSymbol = SymbolFactory.GetAssemblyFromSyntax(syntax, publicKey: _publicKey);
+            IAssemblySymbol rightSymbol = SymbolFactory.GetAssemblyFromSyntax(syntax, publicKey: _publicKey);
 
             // public key tokens must match
             ApiComparer differ = new();
             differ.StrictMode = strictMode;
 
-            IEnumerable<CompatDifference> differences = differ.GetDifferences(leftSymbols, rightSymbols);
```
Comment (Anipik) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/AssemblyIdentityMustMatchTests.cs:168]: can we add a check here that leftAssemblyKeyToken is not null

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/AssemblyIdentityMustMatchTests.cs [hunk 2]
```diff
@@ -202,57 +144,54 @@ public void LeftAssemblyKeyTokenNull(bool strictMode)
...
         }
 
-        [RequiresMSBuildVersionTheory("17.0.0.32901")]
+        [Theory]
         [InlineData(true)]
         [InlineData(false)]
         public void RightAssemblyKeyTokenNull(bool strictMode)
         {
-            var testAsset = _testAssetsManager
-                .CopyTestAsset("PublicKeyTokenValidation", allowCopyIfPresent: true)
```
Comment (Anipik) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/AssemblyIdentityMustMatchTests.cs:211]: can you delete this asset as well from the assets folder

Comment (safern) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/AssemblyIdentityMustMatchTests.cs:211]: Ah, yeah I forgot to do that. Thanks 😄

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/AssemblyIdentityMustMatchTests.cs (hunk 1 above)
Comment (safern) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/AssemblyIdentityMustMatchTests.cs:168]: Yup. I'll add a check in all tests for that kind of information.

**Review by Anipik — APPROVED** [review]
left 2 comments, other than that looks good

**Pull request merged**